### ISSUE TITLE
Documentation popup: \mathcal not resolved/formatted?

### ISSUE THREAD

**Issue by mouse07410** [haskell/haskell-language-server#201]
VS Code with the current `haskell-language-server` master, GHC-8.10.1.

For some reason, when the Documentation popup is displayed - `\mathcal{}` content is not formatted. Here's what the popup shows:
![Screen Shot 2020-07-14 at 11 29 56 ](https://user-images.githubusercontent.com/5923577/87445002-817e7900-c5c5-11ea-8b82-2089bb63fb55.png)

Here's what I see after clicking on "Documentation":
![Screen Shot 2020-07-14 at 10 59 32 ](https://user-images.githubusercontent.com/5923577/87442070-d7512200-c5c1-11ea-98d9-13402fe44f18.png)

Is it normal, or is there a way to make it formatted properly? How can I change my setup/configuration to enable that? Do I need to install some extra package...?

**Comment by ndmitchell** [comment]
The source text actually says mathcal, but then Javascript goes in and replaces it. In general, Javascript won't run on docs being served up from local disk (as its a security risk), which I suspect is what happens. I personally think that relying on Javascript to format a slightly slopey O instead of just doing italics is a dumb case of making everything worse so pedants can have a fractionally more correct O. In Hoogle I just drop the entire mathcal bit, since I'm splicing in snippets of HTML, and don't want some Javascript going on too, which trashes page reflow time etc. See https://github.com/ndmitchell/hoogle/issues/353

**Comment by mouse07410** [comment]
> Javascript won't run on docs being served up **from local disk** (as its a security risk), which I suspect is what happens

As far as I'm concerned, the **real** security risk is a ton of crappy JavaScript pulled from all over the world that a lot of web sites link to. 

I'd much prefer to have a configuration option to allow JavaScript run on docs served from the local disk, as it's an acceptable security risk (after all, I'm running this Cabal executable that came from the 'Net, ain't I? ;). 

However, if it's possible to avoid JavaScript altogether - I'd love to embrace that solution. Is it feasible here?

**Comment by ndmitchell** [comment]
It would be feasible for upstream not to use Javascript, if they wanted faster pages and an irrelevantly less precise font. But upstream seems to have introduced it deliberately and foolishly...

**Comment by mouse07410** [comment]
Ping?

**Comment by mouse07410** [comment]
Ping?

**Comment by ndmitchell** [comment]
Given this issue is fairly minor (visual glitch), and a consequence of upstream making poor decisions, I suspect it's going to be a while before anyone gets around to it. If you're particularly interested, feel free to investigate. For example, is it possible to enable JS in this context? Is JS enabled but the support `.js` file not found? Would it be feasible/desirable to rewrite mathcal before rendering it?

**Comment by OwenGraves** [comment]
I believe this is a similar issue, but it would be nice to have image/gif support on hover as well as math support.
![image](https://user-images.githubusercontent.com/19670245/111577436-4221b100-8780-11eb-8de9-5a1f19da04de.png)
On the haddocks:
![image](https://user-images.githubusercontent.com/19670245/111578468-33d49480-8782-11eb-9a0c-e1f2f02aeabb.png)

Even if displaying math or gifs would be difficult to support, it would be good to at least display the raw text (instead of _cannot render inline math formula_) or the image path location (instead of the little image icon), like what is shown with `:doc`:

![image](https://user-images.githubusercontent.com/19670245/111579034-397eaa00-8783-11eb-927a-63de8953bfda.png)

**Comment by bitwombat** [comment]
Same error shows up in the Vim popup.

**Comment by l8l** [comment]
I still have this problem as of August 2022. I would highly appreciate a fix!

**Comment by refaelsh** [comment]
> I still have this problem as of August 2022. I would highly appreciate a fix!

+1

**Comment by rockboynton** [comment]
I am still seeing this issue as well

**Comment by michaelpj** [comment]
I don't think this is something we can solve very easily. We'd have to return raw HTML for the documentation instead of markdown, I think, after generating images for the latex blocks locally. Perhaps not impossible, but tricky.

**Comment by emlautarom1** [comment]
What about something like https://github.com/fKunstner/latex-to-utf8?

**Comment by webdevred** [comment]
Hi! 👋

I've run into this issue as well. Inline math formulas like \( union t1 t2 \) don't seem to render correctly in Haddock documentation when viewed via ElDoc (in my case, using Emacs with Eglot). It seems that these expressions are not being exported or interpreted by HLS properly.

From what I understand, Haddock relies on JavaScript-based rendering for math expressions, but this isn't included in the documentation data HLS provides, likely due to security concerns around JavaScript.

I'm interested in looking into this more closely when I get the time, and I'd be happy to help investigate or contribute to any solution or workaround.

Thanks for the work on this so far!

**Comment by webdevred** [comment]
Hi again! 👋

I've been experimenting locally with preserving inline and display math from Haddock in haddockToMarkdown, rather than replacing it with the "*cannot render math*" fallback.

Here's what I'm currently doing:

```haskell
haddockToMarkdown (H.DocMathInline s) = "$" ++ s ++ "$"
haddockToMarkdown (H.DocMathDisplay s) = "\n$$\n" ++ s ++ "\n$$\n"
```

This change keeps the original LaTeX content visible in editors like VS Code or Emacs, even if it's not rendered.

I’d like to ask:
- Would this be an acceptable direction for a PR?
- Should we validate or sanitize the LaTeX string to avoid problematic input? For example, should we filter out or reject certain LaTeX commands (\write, \input, \catcode, etc.) for safety?

Happy to iterate on this or contribute a PR if this seems reasonable. Thanks!

Update with additional thought:
Would it make sense to have a user-configurable option in HLS to control how math is rendered in documentation popups?
Option A: Render math expressions as raw LaTeX blocks with `$...$` and `$$...$$` delimiters (current patch approach).
Option B: Parse common LaTeX math expressions and convert them to UTF-8 Unicode symbols where possible (e.g., \mathcal{O} → 𝒪).

Personally, I’d prefer having this as a toggle so users can pick their preferred rendering style depending on their editor or workflow.

Would this be a desirable feature? If yes, I’m happy to help design and implement it.

**Comment by webdevred** [comment]
One more question on this topic:

If support for parsing or transforming LaTeX expressions is something you'd be open to, would you prefer that I:

- Implement a lightweight, custom parser focused on basic math expressions (like` \mathcal{}`,` \frac{}`, etc.), or
- Use an existing library like HaTeX to handle parsing and potentially sanitizing LaTeX input?

The second option would make parsing and transformation easier, but would add a dependency to HLS, so I wanted to check which direction you'd prefer before starting on it.

**Comment by michaelpj** [comment]
> Would this be an acceptable direction for a PR?

That seems pretty reasonable to me. At least showing what's there seems better than just hiding it and saying we can't do anything.

> Should we validate or sanitize the LaTeX string to avoid problematic input?

I think we should at least put it inside a markdown code block. Apart from that, I don't think we need to worry _too_ much: we're not expecting people to be attacked through malicious latex in haddock or anything.

>  Would it make sense to have a user-configurable option in HLS to control how math is rendered in documentation popups?

As much as possible we try to avoid making things configurable if we can avoid it. If we wanted to have a full latex rendering system then that might be heavyweight enough that it needs to be configurable, but ideally not.

Personally I'm a little sceptical that it would be worth the weight, but if we could do it fairly cheaply it might be okay?

**Comment by webdevred** [comment]
Thanks, that makes sense!

I'll go ahead and prepare a PR that just renders the raw LaTeX using `$...$` and `$$...$$`.

I'll skip config options for now, and keep HaTeX out of it unless we decide later that more parsing or transformation is needed.

Thanks for the quick feedback!

**Comment by fendor** [comment]
As there is nothing else HLS can do, I am closing for now
